### PR TITLE
Add support for optional tflint rulesets (Azure, GCP)

### LINKED ISSUE #31: Feature request: ability to install additional plugins (rulesets)
TFLint supports the concept of adding additional rulesets, for instance for the `azurerm` and `google` providers: https://github.com/terraform-linters/tflint#getting-started

To use this action on Terraform configurations that use these providers, an extra option is needed to download and install the appropriate rulesets. For instance:

```yml
with:
  rulesets:
    - azurerm
    - google
```

One would then install that plugin using something like this untested code:
```sh
export TFLINT_PLUGIN_DIR="${TEMP_PATH}/.tflint.d/plugins"
for RULESET in ${INPUT_PLUGINS}; do
  PLUGIN="tf-ruleset-${RULESET}"
  REPOSITORY="https://github.com/terraform-linters/${PLUGIN}"

  echo "::group:: Installing tflint plugin for ${RULESET} ... ${REPOSITORY}"
  mkdir -p ${TFLINT_PLUGIN_DIR}
  curl -L "$(curl -Ls ${REPOSITORY}/releases/latest | grep -o -E "https://.+?_linux_amd64.zip")" -o ${PLUGIN}.zip && unzip ${PLUGIN}.zip -d "${TFLINT_PLUGIN_DIR}" && rm ${PLUGIN}.zip
  echo '::endgroup::'
done
```

### PR DESCRIPTION
**Pull request by Vlaaaaaaad**
**NOTE**: this PR is best reviewed commit-by-commit, as the default file view will be a bit confusing.

For legacy reasons (since it originally only supported AWS), `tflint` ships with all the required rules for [AWS](https://github.com/terraform-linters/tflint-ruleset-aws) pre-packaged. The tflint rules for [Google Cloud Platform](https://github.com/terraform-linters/tflint-ruleset-google) or [Azure](https://github.com/terraform-linters/tflint-ruleset-google) have to be installed manually.

This PR adds optional support for installing those rules. If `tflint_rulesets: google` is set, the [required GCP rules](https://github.com/terraform-linters/tflint-ruleset-google) will be installed and tflint will be configured to use those. There's no actual install script provided, so I had to do some `curl` + `sed` magic to make it a tiny bit portable. 

This PR led to a high number of tests and the GitHub Actions names and configurations were getting confusing. I moved all the test data in a dedicated `test` folder, and did some renaming and moving of GitHub Actions yaml files for more clarity.

Closes https://github.com/reviewdog/action-tflint/issues/31

Note: the expected GitHub Actions checks are due to all my renaming. After the PR is merged, I will change those in the repository settings 🙂

### CODE REVIEW

File: README.md
Comment (haya14busa) [inline README.md]: Add `space separated` so that users can understand how to pass multiple rules.

File: .github/workflows/lint.yml
Comment (haya14busa) [inline .github/workflows/lint.yml:10]: Just out of curiosity, but what is the reason behind this change?

File: script.sh
Comment (haya14busa) [inline script.sh]: 1. It only supports repositories under terraform-linters with certain naming and release names. Should we mention this fact explicitly in the document? I found some plugins under different owners from a quick github search. https://github.com/search?q=tflint-ruleset
2. I think we should check $TFLINT_PLUGIN_DIR is empty or not before overwriting it. Users can actually install custom plugins and set the env var outside of this action and it should work. We may want to write a document about it in addition to the support of INPUT_TFLINT_RULESETS.

**Review by haya14busa — COMMENTED** [review]
LGTM other than nit comments, but note that I'm not familiar with tflint nor its ecosystem ;)

File: .github/workflows/lint.yml
Comment (jsoref) [reply .github/workflows/lint.yml:10]: If you don't have a name, the job itself is used as the name. The other two workflows in the same file use this style. (It looks better...)

Comment (Vlaaaaaaad) [reply .github/workflows/lint.yml:10]: @jsoref is right! I made all the jobs have consistent names 🙂

File: script.sh
Comment (Vlaaaaaaad) [reply script.sh]: Uuuuu, good ideas! I'll implement these 2 before merging this. I did not even think to check what other rulesets are, besides the ones in the official repo 🤦

Comment (Vlaaaaaaad) [reply script.sh]: > It only supports repositories under terraform-linters with certain naming and release names. Should we mention this fact explicitly in the document? I found some plugins under different owners from a quick github search. https://github.com/search?q=tflint-ruleset

Added this to the docs! If there's a need for community rulesets too, we can add that as a follow-up feature.

> I think we should check $TFLINT_PLUGIN_DIR is empty or not before overwriting it. Users can actually install custom plugins and set the env var outside of this action and it should work. We may want to write a document about it in addition to the support of INPUT_TFLINT_RULESETS.

`TFLINT_PLUGIN_DIR` was actually only used by this action (`mktemp -d` + `mkdir -p`), so there was no chance of overwrites happening. 
That said, your initial point was very good! I changed the code to first check if `TFLINT_PLUGIN_DIR` is empty. If it is not, we'll re-use whatever value is set there. This has the added benefit of allowing people to install custom rulesets before this action 🙂 We'll respect that, and reuse the folder!

**Comment by Vlaaaaaaad** [comment]
Hadolint job seems to fail due to https://github.com/reviewdog/action-hadolint/issues/47, even though we are using the `v1` tag. Weird.

**Comment by Vlaaaaaaad** [comment]
@reviewdog/actions-maintainer when y'all have time, a re-review on this would be perfect!

File: .github/workflows/tests.yml
Comment (jsoref) [inline .github/workflows/tests.yml]: Not sure about house style, but personally, I'd do:

```sh
          if [ "$tflint_return" -ne 3 ]; then
            echo "tflint returned ${tflint_return}, expected '3'. Failing..."
            exit 1
          fi
          echo "tflint correctly returned failure ${tflint_return}"
```

File: .github/workflows/test_rulesets.yml
Comment (jsoref) [inline .github/workflows/test_rulesets.yml]: Maybe:
```
if: !cancelled()
```

it's slightly easier to understand than guessing what it's actually doing.

File: script.sh
Comment (jsoref) [inline script.sh]: you might do:
```
uname | tr A-Z a-z
```
instead of using sed s///

File: .github/workflows/test_rulesets.yml
Comment (Vlaaaaaaad) [reply .github/workflows/test_rulesets.yml]: To me, that looks even more confusing: what does `canceled()` even mean (manual cancel? error?)? 

Using `success()` and `failure()` makes it a bit clearer that "_hey, run this step all the time_".

File: script.sh
Comment (Vlaaaaaaad) [reply script.sh]: That could work! My thinking was that I am using `sed` to change the architecture string too, and I'd keep it consistent: people would see the first `sed` and, if they're not already familiar, they would understand it does a replace 🙂

Comment (jsoref) [reply script.sh]: The current code wouldn't work on Windows :-)

Comment (Vlaaaaaaad) [reply script.sh]: Hm... does the `tflint` installer work on Windows? It's named `intall_linux.sh` and has hardcoded non-Windows paths (`/usr`, `/tmp`, `/bin`), but it does have an option for Windows.

Comment (jsoref) [reply script.sh]: 🤷 

The reason I suggest things like `tr` is to avoid hard-coding a list. The fewer specific quirks the less work going forward in case the underlying system supports more things later.

It sounds like GitHub Runners will eventually be using msys2: https://github.com/msys2/setup-msys2#context which I think would include `/usr`, `/tmp`, and `/bin`, so I wouldn't just offhand assume that such things shouldn't work.

Comment (Vlaaaaaaad) [reply script.sh]: I tested the script and I can confirm it works on Windows! I'll review and implement the changes needed to get this also running on Windows

File: script.sh
Comment (jsoref) [inline script.sh]: It looks much nicer this way 😃

Comment (Vlaaaaaaad) [reply script.sh]: You have no idea how many iterations I went through 😆

File: script.sh
Comment (jsoref) [inline script.sh]: ?

**Comment by Vlaaaaaaad** [comment]
Based on the review comments, I pushed a big update! Changes:
- Windows support and tests
- macOS support and tests
- ARM support, with no tests due to GitHub Actions Hosted runners having no ARM support

I changed the `tflint` install process: I copied the `install_linux.sh` script they had and adapted it: instead of installing in `/bin`, it now installs the binary in the same place as reviewdog and the tflint rulesets (a temporary folder to avoid conflicts and permission issues).

There seems to be a weird macOS issue: sometimes GitHub returns a 404 when doing a version check. I have no idea why that happens. A re-run usually fixes it. You can see the same code working at https://github.com/Vlaaaaaaad/action-tflint/pull/2, for example 🤷‍♂️

File: script.sh
Comment (Vlaaaaaaad) [reply script.sh]: The script got rather long, so I added an extra line to separate the big logical chunks: the preparation (dependencies and all that) from actually running `reviewdog`. I think it looks a bit clearer like that.

Comment (jsoref) [reply script.sh]: One blank line I get -- splitting paragraphs
Two blank lines I get -- splitting sections

Three blank lines?

I think that here (seen in a different view w/ more context), you should be removing this blank line because it was associated with:
```
export REVIEWDOG_GITHUB_API_TOKEN="${INPUT_GITHUB_TOKEN}"
```

which you also removed.

File: tests/rulesets/.tflint.hcl
Comment (jsoref) [inline tests/rulesets/.tflint.hcl]: This file is using 4 space indentation. The files above and below it use 2 spaces.

**Comment by Vlaaaaaaad** [comment]
Ping @reviewdog/actions-maintainer for a re-review 🙂

File: script.sh
Comment (haya14busa) [inline script.sh]: This action should not install (overwrite) global reviewdog.
Can you revert this?

Suppose many reviewdog actions install and overwrite reviewdog in global PATH, it could introduce needless complexity and surprise to users.

File: script.sh
Comment (haya14busa) [inline script.sh]: ditto

File: script.sh
Comment (Vlaaaaaaad) [reply script.sh]: Done in https://github.com/reviewdog/action-tflint/pull/35/commits/3306f7cd1ad5e6b59d9554c03c498d73ccdbb835

File: script.sh
Comment (Vlaaaaaaad) [reply script.sh]: Done in https://github.com/reviewdog/action-tflint/pull/35/commits/3306f7cd1ad5e6b59d9554c03c498d73ccdbb835

File: README.md
Comment (haya14busa) [inline README.md:76]: How about mentioning `TFLINT_PLUGIN_DIR` to let users know how to use other custom rulesets by themselves.

**Review by haya14busa — DISMISSED** [review]
LGTM other than the nit comment. :100:

File: README.md
Comment (Vlaaaaaaad) [reply README.md:76]: Good idea! Added!

**Pull request merged**
